File: go/MergeTwoSortedLinkedList/mergeTwoSortedLinkedList.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

// Time -> 0(N) Space -> 0(1)
func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
	if list1 == nil {
		return list2
	}
	if list2 == nil {
		return list1
	}
	head, tail := list1, list2
	if list1.Val > list2.Val {
		head, tail = list2, list1
	}

	merged := head
	mergedHead := merged
	head = head.Next

	for head != nil && tail != nil {
		if head.Val <= tail.Val {
			merged.Next = head
			head = head.Next
		} else {
			merged.Next = tail
			tail = tail.Next
		}
		merged = merged.Next
		merged.Next = nil
	}

	if head == nil && tail != nil {
		merged.Next = tail
	} else if tail == nil && head != nil {
		merged.Next = head
	}
	return mergedHead
}

File: python/mergeTwoSortedLinkedList.py
class ListNode:
    def __init__(self, val=0, next=None) -> None:
        self.val = val
        self.next = next


def mergeTwoLists(list1, list2) -> ListNode:
    if list1 is None:
        return list2
    if list2 is None:
        return list1

    head, tail = list1, list2
    if head.val > tail.val:
        head, tail = tail, head

    merged = head
    mergedHead = merged
    head = head.next

    while head is not None and tail is not None:
        if head.val <= tail.val:
            merged.next = head
            head = head.next
        else:
            merged.next = tail
            tail = tail.next
        merged = merged.next
        merged.next = None

    if head is None and tail is not None:
        merged.next = tail
    elif head is not None and tail is None:
        merged.next = head

    return mergedHead

